nomad/autopilot.go

```diff
@@ -20,88 +22,227 @@ const (
-		conf.RedundancyZoneTag = AutopilotRZTag
-	}
-	if c.EnableCustomUpgrades {
-		conf.UpgradeVersionTag = AutopilotVersionTag
+		Ext:                     autopilotConfigExt(c),
 	}
 
 	return conf
 }
 
-func (d *AutopilotDelegate) FetchStats(ctx context.Context, servers []serf.Member) map[string]*autopilot.ServerStats {
+// FetchServerStats will be called by autopilot to request Nomad fetch the
+// server stats out of band. This method is required to implement the
+// ApplicationIntegration interface
+func (d *AutopilotDelegate) FetchServerStats(ctx context.Context, servers map[raft.ServerID]*autopilot.Server) map[raft.ServerID]*autopilot.ServerStats {
 	return d.server.statsFetcher.Fetch(ctx, servers)
 }
 
-func (d *AutopilotDelegate) IsServer(m serf.Member) (*autopilot.ServerInfo, error) {
-	ok, parts := isNomadServer(m)
-	if !ok || parts.Region != d.server.Region() {
-		return nil, nil
-	}
-
-	server := &autopilot.ServerInfo{
-		Name:   m.Name,
-		ID:     parts.ID,
-		Addr:   parts.Addr,
-		Build:  parts.Build,
-		Status: m.Status,
-	}
-	return server, nil
+// KnownServers will be called by autopilot to request the list of servers known
+// to Nomad. This method is required to implement the ApplicationIntegration
+// interface
+func (d *AutopilotDelegate) KnownServers() map[raft.ServerID]*autopilot.Server {
+	return d.server.autopilotServers()
 }
 
-// NotifyHealth heartbeats a metric for monitoring if we're the leader.
-func (d *AutopilotDelegate) NotifyHealth(health autopilot.OperatorHealthReply) {
+// NotifyState will be called when the autopilot state is updated. The Nomad
+// leader heartbeats a metric for monitoring based on this information. This
+// method is required to implement the ApplicationIntegration interface
+func (d *AutopilotDelegate) NotifyState(state *autopilot.State) {
 	if d.server.raft.State() == raft.Leader {
-		metrics.SetGauge([]string{"nomad", "autopilot", "failure_tolerance"}, float32(health.FailureTolerance))
-		if health.Healthy {
+		metrics.SetGauge([]string{"nomad", "autopilot", "failure_tolerance"}, float32(state.FailureTolerance))
+		if state.Healthy {
 			metrics.SetGauge([]string{"nomad", "autopilot", "healthy"}, 1)
 		} else {
 			metrics.SetGauge([]string{"nomad", "autopilot", "healthy"}, 0)
 		}
 	}
 }
 
-func (d *AutopilotDelegate) PromoteNonVoters(conf *autopilot.Config, health autopilot.OperatorHealthReply) ([]raft.Server, error) {
-	future := d.server.raft.GetConfiguration()
-	if err := future.Error(); err != nil {
-		return nil, fmt.Errorf("failed to get raft configuration: %v", err)
+// RemoveFailedServer will be called by autopilot to notify Nomad to remove the
+// server in a failed state. This method is required to implement the
+// ApplicationIntegration interface. (Note this is expected to return
+// immediately so we'll spawn a goroutine for it.)
+func (d *AutopilotDelegate) RemoveFailedServer(failedSrv *autopilot.Server) {
+	go func() {
+		err := d.server.RemoveFailedNode(failedSrv.Name)
```

Comment:
Is `Name` here the [agent's `name`](https://www.nomadproject.io/docs/configuration#name)? Would it be possible to remove by ID instead in case the cluster has servers with duplicate names (which I _think_ is allow, though not a good idea 😅)?

Reply:
It's the `serf.Member.Name`, which I think is the same thing. I originally had this by ID (which is what Consul does) but our existing `RemoveFailedNode` code only accepts the name and not the ID. I didn't want to change the behavior here.